src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSourceUnitTest.java

```diff
@@ -239,11 +246,54 @@ public void testIntervals() throws IOException {
         }
 
         ReadsDataSource bam2 = new ReadsDataSource(IOUtils.getPath(bam), samReaderFactory);
-        bam2.setTraversalBounds(intervals);
         List<GATKRead> records = Lists.newArrayList();
         for ( GATKRead read : bam2 ) {
             records.add(read);
         }
         return ctx.parallelize(records);
     }
+
+    @DataProvider(name="readPairsAndPartitions")
+    public Object[][] readPairsAndPartitions() {
+        return new Object[][] {
+                { 1, 1, new int[] {2} },
+                { 2, 2, new int[] {2, 2} },
+                { 3, 2, new int[] {4, 2} },
+                { 3, 3, new int[] {2, 2, 2} },
+                { 6, 2, new int[] {6, 6} },
+                { 6, 3, new int[] {4, 4, 4} },
+                { 6, 4, new int[] {4, 2, 4, 2} },
+        };
+    }
+
+    @Test(dataProvider = "readPairsAndPartitions")
+    public void testPutPairsInSamePartition(int numPairs, int numPartitions, int[] expectedReadsPerPartition) throws IOException {
+        JavaSparkContext ctx = SparkContextFactory.getTestSparkContext();
+        SAMFileHeader header = ArtificialReadUtils.createArtificialSamHeader();
+        header.setSortOrder(SAMFileHeader.SortOrder.queryname);
+        JavaRDD<GATKRead> reads = createReads(ctx, header, numPairs, numPartitions);
+        ReadsSparkSource readsSparkSource = new ReadsSparkSource(ctx);
+        JavaRDD<GATKRead> pairedReads = readsSparkSource.putPairsInSamePartition(header, reads);
+        List<List<GATKRead>> partitions = pairedReads.mapPartitions((FlatMapFunction<Iterator<GATKRead>, List<GATKRead>>) it ->
+                Iterators.singletonIterator(Lists.newArrayList(it))).collect();
+        assertEquals(partitions.size(), numPartitions);
+        for (int i = 0; i < numPartitions; i++) {
+            assertEquals(partitions.get(i).size(), expectedReadsPerPartition[i]);
+        }
+        assertEquals(Arrays.stream(expectedReadsPerPartition).sum(), numPairs * 2);
+    }
+
+    private JavaRDD<GATKRead> createReads(JavaSparkContext ctx, SAMFileHeader header, int numPairs, int numPartitions) {
+        final int readSize = 151;
+        final int fragmentLen = 400;
+        final String templateName = "readpair";
+        int leftStart = 10000;
+        List<GATKRead> reads = new ArrayList<>();
+        for (int i = 0; i < numPairs;i++) {
+            leftStart += readSize * 2;
+            int rightStart = leftStart + fragmentLen - readSize;
+            reads.addAll(ArtificialReadUtils.createPair(header, templateName + i, readSize, leftStart, rightStart, true, false));
```

Comment:
Would be good to add test cases with: 1. secondary alignments in addition to primary alignments  2. supplementary alignments in addition to primary alignments, and possibly 3. the case of more than two reads in a fragment (in which case both "first of pair" and "second of pair" can be set).